lib/js-file.js

```diff
@@ -543,7 +556,236 @@ JsFile.prototype = {
         });
 
         return lines;
+    },
+
+    /**
+     * Renders JS-file sources using token list.
+     *
+     * @returns {String}
+     */
+    render: function() {
+        var result = '';
+
+        // For-loop for maximal speed.
+        for (var i = 0; i < this._tokens.length; i++) {
+            var token = this._tokens[i];
+
+            result += token.whitespaceBefore;
+
+            switch (token.type) {
+                // Line-comment: // ...
+                case 'Line':
+                    result += '//' + token.value;
+                    break;
+
+                // Block-comment: /* ... */
+                case 'Block':
+                    result += '/*' + token.value + '*/';
+                    break;
+
+                default:
+                    result += token.value;
+            }
+        }
+        return result;
+    },
+
+    /**
+     * Builds token list using both code tokens and comment-tokens.
+     *
+     * @returns {Object[]}
+     * @private
+     */
+    _buildTokenList: function(codeTokens, commentTokens) {
+        var result = [];
+        var codeQueue = codeTokens.concat();
+        var commentQueue = commentTokens.concat();
+        while (codeQueue.length > 0 || commentQueue.length > 0) {
+            if (codeQueue.length > 0 && (!commentQueue.length || commentQueue[0].range[0] > codeQueue[0].range[0])) {
+                result.push(codeQueue.shift());
+            } else {
+                var commentToken = commentQueue.shift();
+                commentToken.isComment = true;
+                result.push(commentToken);
+            }
+        }
+        return result;
+    },
+
+    /**
+     * Adds JSCS-specific EOF (end of file) token.
+     *
+     * @private
+     */
+    _addEOFToken: function() {
+        var loc = {
+            line: this._lines.length,
+            column: this._lines[this._lines.length - 1].length
+        };
+        this._tokens.push({
+            type: 'EOF',
+            value: '',
+            range: [this._source.length, this._source.length + 1],
```

Comment:
I've added +1 for `tokenEndPosition` indexes... But I can just ignore EOF there...